### ISSUE TITLE
Imaginary Mu branch

### ISSUE THREAD

**Issue by clarkedavida** [LatticeQCD/SIMULATeQCD#35]
Jishnu implemented in the old code some support for configurations at pure imaginary chemical potential. We should bring it over.

**Comment by clarkedavida** [comment]
Well the smearing test and force test are working. Some things that are left to do:
1. Check the DSlashTest.
2. Right now these tests are implemented as separate `main_`. We should just incorporate them into already existing tests with a flag that turns on/off chemical potential.
3. There seem to be redundant if statements? Should eliminate these.
4. Once everything is passing, we should compare timing with the ParallelGPUCode. If it's a bit slower, we might get some speedup by implementing imaginary mu as a template parameter. If it's about the same speed, I would not bother, since this can slow down compile time.

**Comment by clarkedavida** [comment]
- [ ] Also we need to make multiGPU tests eventually. It's not super important at the moment, since all the runs in the foreseeable future can accommodated by one GPU.

**Comment by clarkedavida** [comment]
Well imaginary chemical potential is merged into the main branch. As it stands, the timing is quite good, resulting in a slowdown of O(0.03%), so we don't need to bother with template parameters.

We should still eliminate redundant if statements, it is just strange/confusing coding, and will also give us a negligible performance boost. Once this is done I will close the issue.

I don't want to bother with combining the tests. These are all run automatically by the test script anyhow.

**Comment by clarkedavida** [comment]
I have also checked the ForceProfiling on an 8^3x4 lattice, and found a slowdown of less than 0.3%. Next I will check the time for a small amount of RHMC using the test, which is a 10^4 lattice.

**Comment by clarkedavida** [comment]
rhmc with 10 steps

after changes: 1m 54s
before changes: 1m 52s

So yeah there is essentially no difference in time. With these three tests, we can safely conclude the performance was negligibly impacted.

**Comment by clarkedavida** [comment]
The imaginary mu tests need to be much more stringent.

**Comment by clarkedavida** [comment]
With the fix in  2daa2c953b4a9ea0af39fb14c196e20482a3b3e0, after removing the `if(mu0!=0)` statements, I find that SIMULATeQCD RHMC with `mu0=0` agrees with SIMULATeQCD RHC with `if(mu0!=0)` statements agrees with parallelGPUCode standard RHMC.

I am still not sure what is going on with parallelGPUCode imaginary mu module...

**Comment by clarkedavida** [comment]
parallelGPUCode mu=0.4 does not agree with SIMULATeQCD mu=0.4, which I guess makes sense given everything else we've discovered so far. SIMULATeQCD version passes RHMC test with 90% acceptance and passes a detailed reversibility test.

i will try to implement the more detailed reversibility test in parallelgpucode and see how that goes

**Comment by clarkedavida** [comment]
parallelgpu code with imaginary mu passes the more detailed reversibility test as well as the RHMC with 87% acceptance...

**Comment by clarkedavida** [comment]
It appears to be an issue of precision. In particular in the parallelgpucode normal RHMC and in SIMULATeQCD, there are some hard-coded doubles in integrator.cu. Whereas in the parallelgpucode imaginary mu RHMC these are rather floatT, which then evaluate to float when the tests are run.

**Comment by clarkedavida** [comment]
- [ ] Don't forget documentation

**Comment by clarkedavida** [comment]
I don't really have any interest in getting the imaginary mu to work for multiGPU at the moment, since we will not need it for quite a long time. When I'm done merging in my branch I am going to close this issue.

**Comment by lukas-mazur** [comment]
Isn't this just an extension of the rhmc? I would expect this to work already for multiGPU ?

**Comment by clarkedavida** [comment]
I would also expect it to pass, but now that you mention it, the rhmc is not tested for multiGPU in the current testing script.

**Comment by clarkedavida** [comment]
Imaginary mu stuff is now merged it.
